[TN0010] add response cache eviction solution (#176)
apollographql/docs: src/content/technotes/TN0010-response-cache-eviction.mdx

---
title: Response Cache Eviction
id: TN0010
tags: [server, caching, performance]
---

_For a runnable example of this cache eviction solution, see the [Response Cache Eviction](https://github.com/apollosolutions/response-cache-eviction) repo._

Apollo Server’s Full Response Cache plugin (`apollo-server-plugin-response-cache`) caches the results of operations for a period of time (time-to-live or TTL). After that time expires, the results are evicted from the cache and the server will execute the operation again.

> In practice, this usually means setting a short TTL so that results don’t become stale. 

Reviewer: ```suggestion
The most straightforward way to avoid stale data in the response cache is to set a short default TTL. However, this limits the cache's effectiveness for responses that rarely (or never) change.
```

> As of version `3.7.0`, the Full Response Cache plugin allows customizing cache keys to support new cache eviction patterns. This means it’s easier to use long cache TTLs and increase the hit rate on your cache because you can evict the cached responses when relevant events occur.

Reviewer 2: ```suggestion
As of version `3.7.0`, the Full Response Cache plugin (`apollo-server-plugin-response-cache`) allows customizing cache keys to support new cache eviction patterns. This means it’s easier to use longer cache TTLs and increase the hit rate on your cache by evicting the cached responses when relevant events occur.
```

> As of version `3.7.0`, the Full Response Cache plugin allows customizing cache keys to support new cache eviction patterns. This means it’s easier to use long cache TTLs and increase the hit rate on your cache because you can evict the cached responses when relevant events occur.

Reviewer: ```suggestion
In Apollo Server `3.7.0` and later, the Full Response Cache plugin supports advanced cache eviction patterns via custom cache keys. This enables you to set a longer default TTL and increase the cache's hit rate, because you can selectively evict cached responses when relevant events occur.
```

> ## Customizing the Cache Key

Reviewer: ```suggestion
## Customizing the cache key
```

> This whole solutions works by defining a custom response cache key by a pattern that can later be searched on in the cache. What this key should be comprised of and how it should be structured depends on the use-case and what search patterns our cache implementation supports. 

Reviewer 2: ```suggestion
This works by defining a custom response cache key by a pattern that can later be searched on in the cache. What this key should be comprised of and how it should be structured depends on the use-case and what search patterns our cache implementation supports. 
```

> ### A Note on Cache Key Uniqueness

Reviewer: ```suggestion
### Ensuring cache key uniqueness
```

> Keep in mind that each key links to a full response object, so if your key is too generic, you risk potentially returning the wrong data for queries. I.E, generating a cache key based solely on the operation name would yield the same responses for all operations with the same name, even if the entire query is different. Make sure your keys are unique for each execution of the incoming operations that returns different data. 

Reviewer 2: ```suggestion
Keep in mind that each key links to a full response object, so if your key is too generic, you risk potentially returning the wrong data for queries. For example, generating a cache key based solely on the operation name would yield the same responses for all operations with the same name, even if the entire query is different. Make sure your keys are unique for each execution of the incoming operations that returns different data. 
```

> Keep in mind that each key links to a full response object, so if your key is too generic, you risk potentially returning the wrong data for queries. I.E, generating a cache key based solely on the operation name would yield the same responses for all operations with the same name, even if the entire query is different. Make sure your keys are unique for each execution of the incoming operations that returns different data. 

Reviewer: ```suggestion
Each key in your response cache corresponds to a _complete operation response_. If your method for generating cache keys results in a collision between two unrelated operations, you risk returning the _wrong_ cached response for a particular query.

For example, _do not_ generate cache keys based _solely_ on an operation's name. It's possible that two different clients will execute two different operations with the same name. 
```

### Defining a Custom Cache Key

> Version `3.7.0` of the `apollo-server-plugin-response-cache` introduced the `generateCacheKey` configuration method. Whatever is returned from this function will be used as the cache key to store the current query response.

Reviewer 2: ```suggestion
As noted above, the `3.7.0` of the Full Response Cache plugin introduced the `generateCacheKey` configuration method. The response from this function will be used as the cache key to store the current query response.
```

Here’s the method signature:

```typescript
generateCacheKey(
  requestContext: GraphQLRequestContext<Record<string, any>>,
  keyData: unknown,
): string;

```

> The [requestContext](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-types/src/index.ts#L135-L175) parameter holds data about the running GraphQL request, like the request / response objects as well as the [context object](https://www.apollographql.com/docs/apollo-server/data/resolvers/#the-context-argument) that is passed to your resolver functions. Any portion of these data objects can be used as part of your cache key. 

Reviewer 2: ```suggestion
The [requestContext](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-types/src/index.ts#L135-L175) parameter holds data about the running GraphQL request, such as the request / response objects as well as the [context object](https://www.apollographql.com/docs/apollo-server/data/resolvers/#the-context-argument) that is passed to your resolver functions. Any portion of these data objects can be used as part of your cache key. 
```

> The `keyData` parameter can be used to ensure the uniqueness of your key, simply hashing this variable should be enough to generate a unique key per operation. In fact, the [default implementation](https://github.com/apollographql/apollo-server/blob/76caff6ce797e00907d6ed8e2dab870e255340fe/packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts#L168-L169) just hashes a `JSON.strigify` version of this parameter as the cache key. 

Reviewer: ```suggestion
The `keyData` parameter can be used to ensure the uniqueness of your key. In most cases, hashing this variable should be enough to generate a unique key per operation. In fact, the [default implementation](https://github.com/apollographql/apollo-server/blob/76caff6ce797e00907d6ed8e2dab870e255340fe/packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts#L168-L169) hashes a `JSON.stringify` version of this parameter as the cache key. 
```

In this example, we prefix the default key with the name of the incoming operation:

```typescript
import { createHash } from 'crypto';

function sha(s: string) {
  return createHash('sha256').update(s).digest('hex');
}

generateCacheKey(requestContext, keyData) {
  const operationName = requestContext.request.operationName ?? 'unnamed';
  const key = operationName + ':' + sha(JSON.stringify(keyData));
  return key;
}
```

An example key for the named operation “MyOpName”: 

```bash
keyv:fqc:MyOpName:e7eed80930547ed4ab4ece81a18955967831ff4c40757eda9bf1f0de84e042f8
```

> This approach ensures that all cache keys are unique enough to store unique responses, but gives us a pattern we can leverage to selectively remove cache entries based on our operation names. 

Reviewer: ```suggestion
This approach ensures that all cache keys are unique enough to store unique responses, but gives us a pattern we can use to selectively remove cache entries based on our operation names. 
```

> ## Evicting Cache Entries

Reviewer: ```suggestion
## Evicting cache entries
```

> There are two main strategies for evicting response cache entries: Manually evicting from a shell prompt, or in response to some event, like a mutation. Actually removing entries from the cache once a custom cache key is being used will depend on your caching backend. Different caching tools offer different ways to list and remove keys. We’ll explore both options using Redis. 

Reviewer 2: ```suggestion
There are two main strategies for evicting response cache entries: manually evicting from a shell prompt, or in response to some event, like a mutation. 

Actually removing entries from the cache once a custom cache key is being used will depend on your caching backend, as each offer different ways to list and remove keys. We’ll explore both options using Redis. 
```

> There are two main strategies for evicting response cache entries: Manually evicting from a shell prompt, or in response to some event, like a mutation. Actually removing entries from the cache once a custom cache key is being used will depend on your caching backend. Different caching tools offer different ways to list and remove keys. We’ll explore both options using Redis. 

Reviewer: ```suggestion
There are two primary strategies for evicting response cache entries:

* Manually evicting entries from a shell prompt (usually for debugging or testing)
* Evicting entries automatically in response to some event, such as a mutation

The actual method you use to remove entries depends on your caching backend. Different caching tools offer different ways to list and remove keys. We’ll explore both strategies using Redis. 
```

> ### A Manual Solution

Reviewer: ```suggestion
### Evicting manually
```

> It may be sufficient for your use-case to only need to evict portions of the cache in one-off scenarios, ie, debugging or testing. In such a scenario, simply having a custom cache key pattern in place and shell access to your Redis instance could be enough. 

Reviewer: ```suggestion
If you need to evict cache entries for local testing or debugging, it might suffice to define a custom cache key pattern and delete entries as needed with `redis-cli`.
```

Here’s an example of removing all keys in a Redis instance that match a given pattern:

```bash
redis-cli --raw KEYS "$PATTERN" | xargs redis-cli del
```

> This command will list out every key matching any glob-style "$PATTERN" and remove them one by one. 

Reviewer: ```suggestion
This command lists every key matching any glob-style "$PATTERN" and removes them one by one. 
```

Here’s an example using the operation name prefix pattern described above to remove all entries with unnamed operations:

```bash
redis-cli --raw KEYS "keyv:fqc:unnamed*" | xargs redis-cli del
```

> [The Redis docs for the KEYS command](https://redis.io/commands/keys/) recommend NOT using the `KEYS` function in production application code and only execute against production with “extreme care”. Redis specifically recommends using `SCAN` instead for this type of thing, which is described in the Event Based Solution section.

Reviewer 2: ```suggestion
[The Redis docs for the KEYS command](https://redis.io/commands/keys/) recommend NOT using the `KEYS` function in production application code and only execute against production with “extreme care”. Redis specifically recommends using `SCAN` instead for this use-case, which is described in the Event Based Solution section.
```

Reviewer 2: not 100% this is the right phrasing on my part

> [The Redis docs for the KEYS command](https://redis.io/commands/keys/) recommend NOT using the `KEYS` function in production application code and only execute against production with “extreme care”. Redis specifically recommends using `SCAN` instead for this type of thing, which is described in the Event Based Solution section.

Reviewer: ```suggestion
[The Redis docs for the KEYS command](https://redis.io/commands/keys/) recommend NOT using the `KEYS` function in production application code and only executing against production with "extreme care". Redis specifically recommends instead using `SCAN`, which is described in [Event-based eviction](#event-based-eviction).
```

> The utility of this approach will depend on the number of records stored in your cache and how performant the pattern search is. It will also depend on the number of records that need to be removed. If your searches are scanning and/or returning millions of records, this approach probably should be avoided in a production environment. 

Reviewer 2: ```suggestion
The utility of this approach will depend on the number of records stored in your cache and how performant the pattern search is, as well as the number of records that need to be removed. If your searches are scanning and/or returning millions of records, this approach probably should be avoided in a production environment. 
```

Reviewer 2: we should probably dissuade in general from running these in production. feels like an easy way to get someone to accidentally run a [bobby-tables](https://xkcd.com/327/) like command.

> ### An Event Based Solution

Reviewer: ```suggestion
### Event-based eviction
```

> Most other use-cases need to evict cache entries in response to certain events. The [Response Cache Eviction](https://github.com/apollosolutions/response-cache-eviction) repo, mentioned above, has a full walk-through of evicting certain operation responses from the cache when a specific mutation is executed.

Reviewer: ```suggestion
Most other use cases need to evict cache entries in response to certain events. The [Response Cache Eviction](https://github.com/apollosolutions/response-cache-eviction) repo provides a full walkthrough of evicting certain operation responses from the cache when a specific mutation is executed.
```

> Redis clients offers no way to batch delete entries based on a pattern (that we’ve found at least). Our event based solution needs to do a similar algorithm: Look up keys by a pattern, then remove those keys. 

Reviewer 2: ```suggestion
Redis clients currently offer no way to batch delete entries based on a pattern. As a result, our event based solution needs to do a similar algorithm: look up keys by a pattern, then remove those keys. 
```

i hate english sometimes, not sure if this needs to be capital or not: https://www.tckpublishing.com/capitalization-after-colons/

The following snippet is from [the repo mentioned above](https://github.com/apollosolutions/response-cache-eviction): 

```tsx
import { createClient, RedisClientType } from 'redis';

const deleteByPrefix = async (prefix: string) => {
  const client = createClient({ url: 'redis://localhost:6379' })
  await client.connect()

  const scanIterator = client.scanIterator({
    MATCH: `keyv:fqc:${prefix}*`,
    COUNT: 2000, 
  })

  let keys = []

  for await (const key of scanIterator) {
    keys.push(key)
  }

  if (keys.length > 0) {

>     client.del(keys)

Reviewer 2: Is this blocking or a promise? 

also: it may be worth noting that high deletion counts should be avoided, or if required, done in another thread/process instead to avoid blocking the execution/response longer than needed.

  }

  return keys
}
```

> This solution uses the `scanIterator` function (which uses the `SCAN` Redis function) to scan through cache entries in a memory efficient way, as opposed to the `KEYS` method mentioned above. The `SCAN` method is more appropriate to use in a production environment. 

Reviewer: ```suggestion
This solution uses the `scanIterator` function (which uses the `SCAN` Redis function) to scan through cache entries in a memory-efficient way, as opposed to the `KEYS` method mentioned above. The `SCAN` method is more appropriate to use in a production environment. 
```

The `deleteByPrefix` method can be added to your context object and then executed in your mutation resolvers to remove certain operations from the cache. 

> Again a full working example of this is available in the [Response Cache Eviction](https://github.com/apollosolutions/response-cache-eviction) repo.

Reviewer: ```suggestion
```

> ## Final Thoughts

Reviewer: ```suggestion
## Final thoughts
```

> Either of the eviction solutions mentioned above should be used with caution. Make sure you have an understanding of the sorts of effects that your set up will have on your cache. Its a good idea to monitor your caching server when testing our your different use-cases to ensure that you aren’t overloading your cache.

Reviewer: ```suggestion
Either of the eviction solutions mentioned above should be used with caution. Make sure you have an understanding of the sorts of effects that your setup will have on your cache. Its a good idea to monitor your caching server when testing your different use cases to ensure that you aren’t overloading your cache.
```

tags: [server, caching, performance]
---

_For a runnable example of this cache eviction solution, see the [Response Cache Eviction](https://github.com/apollosolutions/response-cache-eviction) repo._

Apollo Server’s Full Response Cache plugin (`apollo-server-plugin-response-cache`) caches the results of operations for a period of time (time-to-live or TTL). After that time expires, the results are evicted from the cache and the server will execute the operation again.

In practice, this usually means setting a short TTL so that results don’t become stale. 

As of version `3.7.0`, the Full Response Cache plugin allows customizing cache keys to support new cache eviction patterns. This means it’s easier to use long cache TTLs and increase the hit rate on your cache because you can evict the cached responses when relevant events occur.

## Customizing the Cache Key

This whole solutions works by defining a custom response cache key by a pattern that can later be searched on in the cache. What this key should be comprised of and how it should be structured depends on the use-case and what search patterns our cache implementation supports. 

### A Note on Cache Key Uniqueness

> Keep in mind that each key links to a full response object, so if your key is too generic, you risk potentially returning the wrong data for queries. I.E, generating a cache key based solely on the operation name would yield the same responses for all operations with the same name, even if the entire query is different. 
> 
> Put simply, make sure your keys are unique enough for each incoming operation that returns different data. 

Reviewer 3: ```suggestion
Keep in mind that each key links to a full response object, so if your key is too generic, you risk potentially returning the wrong data for queries. I.E, generating a cache key based solely on the operation name would yield the same responses for all operations with the same name, even if the entire query is different. Make sure your keys are unique for each execution of the incoming operations that returns different data. 
```

### Defining a Custom Cache Key

> Version `3.7.0` of the `apollo-server-plugin-response-cache` introduced the `generateCacheKey` configuration method. Whatever is returned from this function will be used as the cache key to store the current query response

Reviewer 3: ```suggestion
Version `3.7.0` of the `apollo-server-plugin-response-cache` introduced the `generateCacheKey` configuration method. Whatever is returned from this function will be used as the cache key to store the current query response.
```

```

The [requestContext](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-types/src/index.ts#L135-L175) parameter holds data about the running GraphQL request, like the request / response objects as well as the [context object](https://www.apollographql.com/docs/apollo-server/data/resolvers/#the-context-argument) that is passed to your resolver functions. Any portion of these data objects can be used as part of your cache key. 

The `keyData` parameter can be used to ensure the uniqueness of your key, simply hashing this variable should be enough to generate a unique key per operation. In fact, the [default implementation](https://github.com/apollographql/apollo-server/blob/76caff6ce797e00907d6ed8e2dab870e255340fe/packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts#L168-L169) just hashes a `JSON.strigify` version of this parameter as the cache key. 

In this example, we prefix the default key with the name of the incoming operation:

```typescript
import { createHash } from 'crypto';

function sha(s: string) {
  return createHash('sha256').update(s).digest('hex');
}

generateCacheKey(requestContext, keyData) {

>   const operationName = requestContext.request.operationName ?? 'unnamed'
>   const key = operationName + ':' + sha(JSON.stringify(keyData))
>   return key

Reviewer 3: ```suggestion
  const operationName = requestContext.request.operationName ?? 'unnamed';
  const key = operationName + ':' + sha(JSON.stringify(keyData));
  return key;
```
nit: other functions have `;`s

Author: I've got a prettier.rc setting that removes all the unnecessary `;` in this repo

## Evicting Cache Entries

There are two main strategies for evicting response cache entries: Manually evicting from a shell prompt, or in response to some event, like a mutation. Actually removing entries from the cache once a custom cache key is being used will depend on your caching backend. Different caching tools offer different ways to list and remove keys. We’ll explore both options using Redis. 

### A Manual Solution

It may be sufficient for your use-case to only need to evict portions of the cache in one-off scenarios, ie, debugging or testing. In such a scenario, simply having a custom cache key pattern in place and shell access to your Redis instance could be enough. 

Here’s an example of removing all keys in a Redis instance that match a given pattern:

```bash
redis-cli --raw KEYS "$PATTERN" | xargs redis-cli del
```

This command will list out every key matching any glob-style "$PATTERN" and remove them one by one. 

Here’s an example using the operation name prefix pattern described above to remove all entries with unnamed operations:

```bash
redis-cli --raw KEYS "keyv:fqc:unnamed*" | xargs redis-cli del
```

> [The Redis docs for the KEYS command](https://redis.io/commands/keys/) recommend NOT using the `KEYS` function in production application code and only execute against production with “extreme care”. Redis specifically recommends using `SCAN` instead for this type of thing, which is described in the Event Based Solution section

Reviewer 3: ```suggestion
[The Redis docs for the KEYS command](https://redis.io/commands/keys/) recommend NOT using the `KEYS` function in production application code and only execute against production with “extreme care”. Redis specifically recommends using `SCAN` instead for this type of thing, which is described in the Event Based Solution section.
```

The following snippet is from [the repo mentioned above](https://github.com/apollosolutions/response-cache-eviction): 

```tsx
import { createClient, RedisClientType } from 'redis';

const deleteByPrefix = async (prefix: string) => {
  const client = createClient({ url: 'redis://localhost:6379' })
  await client.connect()

  const scanIterator = client.scanIterator({
    MATCH: `keyv:fqc:${prefix}*`,
    COUNT: 2000, 
  })

  let keys = []

  for await (const key of scanIterator) {
    keys.push(key)
  }

  if (keys.length > 0) {
    client.del(keys)
  }

  return keys
}
```

This solution uses the `scanIterator` function (which uses the `SCAN` Redis function) to scan through cache entries in a memory efficient way, as opposed to the `KEYS` method mentioned above. The `SCAN` method is more appropriate to use in a production environment. 

The `deleteByPrefix` method can be added to your context object and then executed in your mutation resolvers to remove certain operations from the cache. 

> Again a full working example of this is available in the [Response Cache Eviction](https://github.com/apollosolutions/response-cache-eviction) repo

Reviewer 3: ```suggestion
Again a full working example of this is available in the [Response Cache Eviction](https://github.com/apollosolutions/response-cache-eviction) repo.
```
